proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -24,77 +24,148 @@ vector/matrix operations described in [0029].
+* `dx::linalg::MulAdd`
+  * Multiply a matrix in memory by a vector parameter, and add a vector from
+    memory.
+* `dx::linalg::OuterProductAccumulate`
+  * Compute the outer product of two vectors and accumulate the result matrix
+    atomically-elementwise in memory.
+* `dx::linalg::VectorAccumulate`
+  * Accumulate elements of a vector atomically-elementwise to corresponding
+    elements in memory.
+* `dx::linalg::InterpretedVector`
+  * Convenience function to construct a `Vector` inline while inferring various
+    template parameters.
+
+
+These are all described in more detail below, but the follow code example gives
+a flavor of how these work together:
 
 ```c++
-ByteAddressBuffer inputMatrix0; 
-ByteAddressBuffer inputMatrix1; 
-ByteAddressBuffer biasVector0; 
-ByteAddressBuffer biasVector1;
+ByteAddressBuffer model;
+
+vector<float, 3> ApplyNeuralMaterial(vector<float, 8> inputVector) {
+    using namespace dx::linalg;
 
-void ps_main(args) // args: texture, normal, position
-{   
-    PreProcessing(args);
-    // Neural Network computes the output vector
-    // using the same input args and trained data
-    // in the form of matrices and bias vectors.
+    MatrixRef<
+        DATA_TYPE_E4M3, 8, 32, 
+        MATRIX_LAYOUT_INFERENCING_OPTIMAL> matrix0 = { model, 0 };
 
-    // The input vector is computed from the shader input
-    vector<uint32_t, M> inputVector = SomeFunction(args);
+    VectorRef<DATA_TYPE_E4M3> biasVector0 = { model, 1024 };
 
-    // Below the physical calculations are replaced by NN evaluation
-    // the Matrix and Bias are trained offline and loaded to memory.
+    MatrixRef<
+        DATA_TYPE_E4M3, 
+        32, 32, MATRIX_LAYOUT_INFERENCING_OPTIMAL> matrix1 = { model, 2048 };
+    
+    VectorRef<DATA_TYPE_E4M3> biasVector1 = { model, 3072 };
 
-    // layer0 = inputVector*inputMatrix + biasVector0
-    // The matrix and bias are loaded from memory at offsets : moffset0 and boffset0
+    MatrixRef<
+        DATA_TYPE_E4M3, 
+        32, 3, MATRIX_LAYOUT_INFERENCING_OPTIMAL> matrix2 = { model, 4096 };
 
-    dx::linalg::MatrixRef inMat0 = {inputMatrix0, moffset0};
-    dx::linalg::VectorRef biasV0 = {biasVector0, boffset0};
-    vector<uint32_t, K> layer0 = dx::linalg::MulAdd(inputVector, inMat0, biasV0);
-    layer0 = max(layer0,0); // Apply activation function
+    VectorRef<DATA_TYPE_E4M3> biasVector2 = { model, 5120 };
 
-    // layer0 = inputVector*inputMatrix0 + biasVector0
-    // The matrix and bias are loaded from memory at offsets : moffset1 and boffset1
 
-    dx::linalg::MatrixRef inMat1 = {inputMatrix1, moffset1};
-    dx::linalg::VectorRef biasV1 = {biasVector1, boffset1};
-    vector<uint32_t, K> layer1 = dx::linalg::MulAdd(layer0, inMat1, biasV1);
-    layer1 = max(layer1,0); // Apply activation function
+    vector<float, 32> layer0 = MulAdd(
+        InterpretedVector<DATA_TYPE_FLOAT16>(inputVector), 
+        matrix0, 
+        biasVector0);
```

Comment:
I get that this is sample code but it seems unlikely that you'd want to:
* Take input float32
* Convert to float16
* Multiply with float8 (E4M3)
* Bias with float8
* Return float32

This also isn't in the guaranteed supported set. You probably want this to be:
* Input float32 (or float16, which is in the minimum required support set)
* Convert to float8
* Multiply with float8
* Bias with float16
* Return float16

Reply:
How's the new one look?

Reply:
Yeah that looks more plausible (based on my complete non-expert understanding of the spec).